src/control/tools/EditSelection.cpp

```diff
@@ -258,6 +259,12 @@ auto EditSelection::setFill(int alphaPen, int alphaHighligther) -> UndoAction* {
     return this->contents->setFill(alphaPen, alphaHighligther);
 }
 
+/**
+ * Set the line style of all elements, return an undo action
+ * (Or nullptr if nothing done)
+ */
+auto EditSelection::setLineStyle(LineStyle style) -> UndoAction* { return this->contents->setLineStyle(style); }
```

Comment:
It is intended to use only smart ptr's. Since the target requires a smart pointer and this is new code; use smart pointers here:

```suggestion
auto EditSelection::setLineStyle(LineStyle style) -> UndoActionPtr { return this->contents->setLineStyle(style); }
```

Reply:
@Febbe 
The compiler complains 
```
error: ‘UndoActionPtr’ does not name a type; did you mean ‘UndoAction’ 
```
Should I duplicate the `using` statement from the UndoRedoHandler class? However there are more errors coming along. Everything compiles when I change back to `UndoAction*` (4 places).

Reply:
This warning means, that the compiler does not know jet what this type is.
To copy the using statement could result into a name conflict, when suddenly both statements occur in the same compilation unit. So the best would be to either use the complete name or just move the usng UndoActionPtr statement into the UndoAction.h and include that.

Reply:
Ok, I moved the `using UndoActionPtr` into the `UndoAction.h` header file. In order to make the line 
`auto undo = std::make_unique<LineStyleUndoAction>(this->sourcePage, this->sourceLayer);` compile I had to move the `LineStyleUndoActionEntry` struct to the header file. Can you please take a look if that was fine?

Reply:
> compile I had to move the LineStyleUndoActionEntry struct to the header file. Can you please take a look if that was fine?

Yes that's fine.
But it was not required (I think so), the reason it can't compile is, that the it is not known, how to destruct the vector in your class, since you only forward declared the LineStyleUndoActionEntry, but you try to delete the vector already in the header file.

To resolve that there are 2 ways: Do not forward declare the class and define it in the header (like you did).
Or move the definition (not the declaration) of the destructor of LineStyleUndoAction to the source file.

LineStyleUndoAction.h: `~LineStyleUndoAction();` <= Rule of 5 is recommended but not mandatory
LineStyleUndoAction.cpp: `~LineStyleUndoAction() = default;` <= Always Requires Rule of 5

Reply:
As I understand everything is fine now as it is, right? So I can merge?

Reply:
Yes, Ship it